### PR TITLE
Add Shard Path Into Slow Log

### PR DESCRIPTION
**Pull request by gf2121**
It is usually necessary to analyze the disk status to find out whether the disk IO bottleneck made the search slow, so I added shard path information into the search slow log to make the troubleshooting easier.

### CODE REVIEW

**Comment by gf2121** [comment]
I have already signed a contributor agreement. Would this refresh automatically? Or I need to open another pull request?

**Comment by garyelephant** [comment]
@jakelandis 
Sorry to disturb, any progress ? I saw this PR weeks ago and I think it is also useful for me. Can this PR be merged into master as soon as possible ?

**Comment by garyelephant** [comment]
@elasticmachine

**Comment by pugnascotia** [comment]
@elasticmachine ok to test

**Comment by pugnascotia** [comment]
@gf2121 can you explain more about how this information would help you diagnose performance issues?

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Comment by pugnascotia** [comment]
@gf2121  This branch needs to be updated from the current `master`

**Comment by pgomulka** [comment]
@gf2121 you can also achieve that by checking the `shardId` which is already present in logs. Then you can find the nodeId/name in `_cat/shards` and with that information you can check the `data paths` in `_cluster/settings`
Let me know if that would work for you

and I also second @pugnascotia on the question about the use case.

**Comment by gf2121** [comment]
> @gf2121 can you explain more about how this information would help you diagnose performance issues?

Shard path tells me which disk the data is allocated to. This information can be correlated with the disk monitoring (such as iowait) at that time for analysis. If a query is slowed down by the disk IO bottleneck, it would be quite easily seen with the shard path information.

**Comment by gf2121** [comment]
> @gf2121 @garyelephant you can also achieve that by checking the `shardId` which is already present in logs. Then you can find the nodeId/name in `_cat/shards` and with that information you can check the `data paths` in `_cluster/settings`
> Let me know if that would work for you
> 
> and I also second @pugnascotia on the question about the use case.

@pgomulka 
1. _cat/shards and _cluster/setting API can indeed obtain the shard allocation path, but if it can be directly obtained from the slow log, it will greatly improve the efficiency of troubleshooting.
2. Record the shard path in log can help keep the scene. For example, if shard path was not recorded and a shard was relocated after the slow query occored, I cannot know where the shard was stored (real shard path at the time slow query occored) when I check the slow log, and even the current shard path gotten by APIs will mislead me.

**Comment by gf2121** [comment]
@elasticmachine update branch

**Comment by gf2121** [comment]
> @gf2121 This branch needs to be updated from the current `master`

@pugnascotia Thanks, i have updated it.

**Comment by gf2121** [comment]
@pugnascotia @pgomulka Is my explanation clear? or if there is any other questions, just tell me.

**Comment by pgomulka** [comment]
@gf2121 I understand your reasoning, however I still think we should not include `shard path` in search slow logs. These are used in many different workflows and we have be careful to not overload it with too much information.
Your use case looks a bit too narrow and specific to be include in slow logs which are used by everyone. I suspect that you use multiple data paths in your cluster, maybe you can consider not using that feature to make your investigation simpler?

This is true that information in logs might be stale when investigating the problem. This applies to all logs not only the search slow logs we are discussing here.

We really appreciate your effort though and we will happily review your future PR or discuss concerns on https://discuss.elastic.co/

For the reasons above I will close the PR

**Comment by gf2121** [comment]
@pgomulka  I am so sorry for my poor English, let me give a more specific example here, where my cluster has 2 nodes: 

Node1 : `1.1.1.1(0)`  mounted three disks: `/data00` `/data01` and `/data02`.
Node2 : `1.1.1.2(0)`  mounted three disks: `/data00` `/data01` and `/data02`.

and a index:

Index: myIndex with one shard `myindex[0]`

Here is the story:

**2:00 a.m.** A slow search comes, `myindex[0]` is in `1.1.1.1(0)`.  a slow log is recorded. 

**4:00 a.m.**  `myindex[0]` is relocated to `1.1.1.2(0)`.

**6:00 a.m.** I check the slow log for reason , and find that `myindex[0]` was in `1.1.1.1(0)`, but now, the shard is in `1.1.1.2(0)`. So I can not tell which disk the shard was in at 2:00 a.m.

In a word, I am not complaining that log is stale, instead, a "stale" log is just what I want, because it can tell me the true path at 2:00 a.m.

**Comment by gf2121** [comment]
`shard path` is usually short , so i don't think it will make the slow log overload.

**Comment by gf2121** [comment]
@pgomulka  Shard path is usually short , so I don't think it will make the slow log overload.
We indeed use multiple data path in a node,  we have built a lot of ES clusters in the production environment, and each cluster maintains PB-level data. We think using multiple data path in a node is a very common feature. For example, if I have a server whose memory capacity can just build one node, but the server has 2 disks, we will not give up one disk just to avoid this feature.

In summary, I really think that shard path is essential in slow log, I hope you can think about it again, thank you very much~

**Comment by pgomulka** [comment]
@gf2121 I raised this for a discussion again within the team. The conclusion is that we don't think shardId is needed in slow logs. 
We understand your usecase, it is tempting to use slow logs to identify hardware issues, but this was not its primary purpose. Slow logs are meant to help you identify which a particular query or ingestion was slow. 
It is true that shardId is not long, but it just won't fit there. There is a lot of other fields available within that context that we don't include for the same reason. We want slow logs to be concise and aim to help identifying slow queries.

As per your usecase I think you need to look into options outside of Elasticsearch. Consider using scripts that would possibly poll the cluster settings which would give you shardId /shardPath information. You can then correlate these with slowlogs to filter down. I think it still won't be enough and you need to use some 3rd party tools like iostat and similar.
Feel free to reach out and ping me on https://discuss.elastic.co/t/log4j-regex-filter-to-avoid-audit-logs/226016
